### ISSUE TITLE
Pods removed but metrics exported

### ISSUE THREAD

**Issue by tomfankhaenel** [jmcgrath207/k8s-ephemeral-storage-metrics#124]
Hi,

the exporter works very nice but i recognized a strange behavior and maybe it is just a configuration issue but i dont know how to fix this:

Scenario:
I got a node where lots of gitlab runner jobs are spawned in lots of different pods which only live for a couple of minutes. The interval of 15s is enough to catch a few metrics of the usage and the metrics are exported as expected.
I was trying to create a grafana dashboard where i can see the usage of the ephemeral storage over all pods . There i recognized a very very hughe usage which was not possible and it was also increasing very fast which should result in lots of disk pressuere.

After looking into the metrics the exporter provided i found lots of pods which already vanished minutes/hours ago. The exporter is acting as if the pod is still running and exporting the last value over and over again.

Example:

![image](https://github.com/user-attachments/assets/fd16cc84-090a-449f-aa09-b79388b61306)

Logs:
`{"level":"info","time":1729849523,"message":"Node ip-10-4-102-250.eu-central-1.compute.internal does not exist. Removed from monitoring"}
{"level":"warn","time":1729849534,"message":"Failed fetched proxy stats from node : ip-10-4-102-250.eu-central-1.compute.internal: the server is currently unable to handle the request"}
{"level":"warn","time":1729849534,"message":"Could not query node ip-10-4-102-250.eu-central-1.compute.internal for ephemeral storage"}
{"level":"warn","time":1729849549,"message":"Failed fetched proxy stats from node : ip-10-4-102-250.eu-central-1.compute.internal: the server is currently unable to handle the request"}
{"level":"warn","time":1729849549,"message":"Could not query node ip-10-4-102-250.eu-central-1.compute.internal for ephemeral storage"}
{"level":"warn","time":1729849656,"message":"Failed fetched proxy stats from node : ip-10-4-85-197.eu-central-1.compute.internal: the server is currently unable to handle the request"}
{"level":"warn","time":1729849656,"message":"Could not query node ip-10-4-85-197.eu-central-1.compute.internal for ephemeral storage"}`

Settings:

```
The current values are used:
# -- Set metrics you want to enable
metrics:
  # -- Percentage of ephemeral storage used by a container in a pod
  ephemeral_storage_container_limit_percentage: false
  # -- Current ephemeral storage used by a container's volume in a pod
  ephemeral_storage_container_volume_usage: true
  # -- Percentage of ephemeral storage used by a container's volume in a pod
  ephemeral_storage_container_volume_limit_percentage: false
  # -- Current ephemeral byte usage of pod
  ephemeral_storage_pod_usage: true
  # -- Available ephemeral storage for a node
  ephemeral_storage_node_available: false
  # -- Capacity of ephemeral storage for a node
  ephemeral_storage_node_capacity: false
  # -- Percentage of ephemeral storage used on a node
  ephemeral_storage_node_percentage: false
  # -- Create the ephemeral_storage_adjusted_polling_rate metrics to report Adjusted Poll Rate in milliseconds. Typically used for testing.
  adjusted_polling_rate: false

resources:
  limits:
    memory: 128Mi
  requests:
    cpu: 50m
    memory: 128Mi

# Pod Security Context
# ref: https://kubernetes.io/docs/tasks/configure-pod-container/security-context/
podSecurityContext:
  runAsNonRoot: true
  runAsUser: 10001
  runAsGroup: 10001
  readOnlyRootFilesystem: true
  seccompProfile:
    type: RuntimeDefault

# Container Security Context to be set on the controller component container
# ref: https://kubernetes.io/docs/tasks/configure-pod-container/security-context/
containerSecurityContext:
  allowPrivilegeEscalation: false
  capabilities:
    drop:
      - ALL
  privileged: false
  readOnlyRootFilesystem: true
  runAsNonRoot: true
```

Is there something wrong in the configuration or why does the exporter does not remove the pods which are no longer running on the nodes? I  use the exporter with an istio-proxy in front and made some slight security adjustements.

**Comment by jmcgrath207** [comment]
Hi @tomfankhaenel,

Thanks for raising this issue. I agree this is not desired behavior as those pods should be evicted and you config looks good to me.

Looking at our e2e test we currently only testing for pod eviction when a node disappears. So I will be adding this test case to reproduce this.

https://github.com/jmcgrath207/k8s-ephemeral-storage-metrics/blob/master/tests/e2e/deployment_test.go#L377-L384

I will follow up once I find out more on this issue.

**Comment by jmcgrath207** [comment]
Hi @tomfankhaenel,

It took a bit longer, but I was able to reproduce your issues with unresponsive nodes.

At a high level, if a delete event was received, the node and all associated pods were removed from the exporter. However, I found there was no handling if the node was still observed by kubernetes (via kubectl) but was in a `Not Ready` status. This caused old pods not to be evicted.

The PR fixes this above, and your change lives in `1.14.0`.

I'm closing the ticket, but feel free to reopen it if it's unresolved. 

Thanks!

**Comment by tomfankhaenel** [comment]
Hi @jmcgrath207,

thank you for the response but it looks my issue was not fixed. From what i understood the changes does not affect the issue above.

The nodes are running and are ready in all cases but the pods are stopped (in my case because the gitlab job finished) but the metric keeps existing forever.


Maybe these screenshots can explain the issue. This is the metric i currently get for a pod which was stopped around 8:56
![image](https://github.com/user-attachments/assets/fb8ae3df-aabc-4417-8047-9e409c538aa1)

As you can see the metric will be exported over and over again
![image](https://github.com/user-attachments/assets/e5e29c69-d82b-4467-a73c-41da66e27859)

I would expect the metrics to be exported only during the lifetime of the pod similar like this:
![image](https://github.com/user-attachments/assets/598ab8c2-2105-41b7-8e9b-5c38f529b63c)

At the time the pod was stopped the node it was running on was still up and running:
![image](https://github.com/user-attachments/assets/d8c31b77-d7e5-45d9-a572-0b8ff4eb94c1)


In short: I expect the ephemeral_storage_pod_usage to be only exported for pods that are in the status running. As soon as a pod is stopped it should be removed from the exporter.
At the moment the pods are only removed in case the exporter is restarted.

I think there is another issue. I stopped the node where the pod was running on:
![image](https://github.com/user-attachments/assets/802b601d-ff82-42f7-ae00-4a41ae2a63e9)

but the exporter is still exporting the usage metric of the pod:
![image](https://github.com/user-attachments/assets/60f76a79-0aab-44f5-8475-0391c3b72b90)


My current workaround is to restart the exporter every 15min. to cleanup obsolete metrics.

**Comment by jmcgrath207** [comment]
Thanks for going into depth here @tomfankhaenel 

I will take a deeper look into the pod eviction logic and reproduce this issue with a chart. Seems the node error message in the first post was a slight red herring.

As for current testing, I just regex scrape or look at the exporter endpoint page with a port forward.  I am hoping this chart approach will surface your error more clearly this time for me 🫰

Heads up, it might be a week until I get another chance to troubleshoot since I am traveling.

**Comment by jmcgrath207** [comment]
Hey @tomfankhaenel, I just got back from vacation and took a deeper look. I observed the pod disappear from the Prometheus graph and exporter page and confirmed that the deleted pod no longer appears as expected. I also scaled out Minikube to delete the pods and saw the same desired behavior.

![image](https://github.com/user-attachments/assets/646edd84-f340-44a5-a6a4-7418b76e4fe6)


Eviction works by removing a node that becomes unresponsive after a couple retires or a delete event is received.

Pod eviction only works if a delete event is received. I believe that is our issue here (see node eviction logic below).
https://github.com/jmcgrath207/k8s-ephemeral-storage-metrics/blob/master/pkg/node/metrics.go#L94

This will require a refactor so that all metrics variables can be shared across modules so eviction can happen regarding the Kubernetes type (e.g., Pods, Nodes). It should only take a day or two.

Once again, thank you for your patience and for bringing this issue to our attention; we will get it fixed soon.

**Comment by jmcgrath207** [comment]
I merged the fix for this in release `1.15.0`. Let me know if it's corrected @tomfankhaenel 

Thanks!

**Comment by tomfankhaenel** [comment]
@jmcgrath207 Thank you for looking into this issue. I can confirm the behavior has improved as now the pods that where running on a pod that has been removed from the cluster are removed as soon as the node is no longer existing:
After the event:
`{"level":"info","time":1732528745,"message":"Node ip-10-4-90-27.eu-central-1.compute.internal does not exist or is unresponsive. Removed from monitoring"}`
we can see the metrics of the pod will be removed from the exporter:
![image](https://github.com/user-attachments/assets/9a053bea-03f8-4ece-8fdb-712720a750c2)

However ... pods that are stopped and removed from the cluster for example on that specific node are still exported forever. You can verify this by the flatlines until 9:59. All these pods where already stopped but metrics still exported.

Is there something wrong with my cluster setup that the pod eviction / deletion is not visible to the exporter?

**Comment by jmcgrath207** [comment]
@tomfankhaenel 

Thanks for retesting this. Nice to hear we are making some progress.


Is it possible you can send me a `kubectl get pods` to show the status of the pod. I am thinking that the pod is in complete status and is not being evicted here.

**Comment by tomfankhaenel** [comment]
The pods are terminated and do not exist after they finished.
![image](https://github.com/user-attachments/assets/7d7b7b3d-4475-4deb-b6a1-08bbb6c60a20)
![image](https://github.com/user-attachments/assets/dc807386-cd65-4616-807b-37120f6f4ad8)
![image](https://github.com/user-attachments/assets/ec3edb12-c8b1-4bd1-b731-ac92288caa1e)

**Comment by tomfankhaenel** [comment]
@jmcgrath207 Could my behavior be another issue or is this expected behavior? As the pods are terminated i was expecting they should be removed from the exporter results.

**Comment by tomfankhaenel** [comment]
I looked into the API as described for a node where a runner was spawned and after a couple of minutes finished and exited as usual.
The result is the runner pod shows up in this call and after it finished the pod is no longer listed in the api call.
The exporter however was exporting the latest metric of this api call over and over again but it was no longer listed. Only after the restart of the exporter itself it was no longer existing. 

It looks to me that the exporter never forgets a pod it saw in this summary:
The content from the api page that is fetched here (https://github.com/jmcgrath207/k8s-ephemeral-storage-metrics/blob/e4c56dee4b8b1fe514d5fc3277c2be90071e9b5e/cmd/app/main.go#L55) is later directly added to the exporter and never questioned again?


This is how the first api call looked like (i removed some other pods for compliance reasons):

```
{
 "node": {
  "nodeName": "ip-10-4-90-51.eu-central-1.compute.internal",
 "pods": [
  {
   "podRef": {
    "name": "runner-km28rwbsd-project-1497-concurrent-5-37qnufcj",
    "namespace": "deployment",
    "uid": "57e6530e-5d3c-4403-88bb-b32670fbabf7"
   },
   "startTime": "2024-12-19T11:28:22Z",
   "containers": [
    {
     "name": "build",
     "startTime": "2024-12-19T11:28:48Z",
     "cpu": {
      "time": "2024-12-19T11:29:57Z",
      "usageNanoCores": 3291484063,
      "usageCoreNanoSeconds": 139133159697
     },
     "memory": {
      "time": "2024-12-19T11:29:57Z",
      "availableBytes": 2875965440,
      "usageBytes": 1424543744,
      "workingSetBytes": 1419001856,
      "rssBytes": 1399799808,
      "pageFaults": 378642,
      "majorPageFaults": 0
     },
     "rootfs": {
      "time": "2024-12-19T11:29:59Z",
      "availableBytes": 76115410944,
      "capacityBytes": 85886742528,
      "usedBytes": 589824,
      "inodesFree": 41766334,
      "inodes": 41942000,
      "inodesUsed": 24
     },
     "logs": {
      "time": "2024-12-19T11:30:04Z",
      "availableBytes": 76115410944,
      "capacityBytes": 85886742528,
      "usedBytes": 4096,
      "inodesFree": 41766334,
      "inodes": 41942000,
      "inodesUsed": 1
     },
     "swap": {
      "time": "2024-12-19T11:30:04Z",
      "swapAvailableBytes": 0,
      "swapUsageBytes": 0
     }
    },
    {
     "name": "helper",
     "startTime": "2024-12-19T11:28:48Z",
     "cpu": {
      "time": "2024-12-19T11:30:03Z",
      "usageNanoCores": 541287,
      "usageCoreNanoSeconds": 15509053920
     },
     "memory": {
      "time": "2024-12-19T11:30:03Z",
      "availableBytes": 2995539968,
      "usageBytes": 2608832512,
      "workingSetBytes": 225685504,
      "rssBytes": 14733312,
      "pageFaults": 65142,
      "majorPageFaults": 0
     },
     "rootfs": {
      "time": "2024-12-19T11:29:59Z",
      "availableBytes": 76115410944,
      "capacityBytes": 85886742528,
      "usedBytes": 972980224,
      "inodesFree": 41766334,
      "inodes": 41942000,
      "inodesUsed": 21
     },
     "logs": {
      "time": "2024-12-19T11:30:04Z",
      "availableBytes": 76115410944,
      "capacityBytes": 85886742528,
      "usedBytes": 4096,
      "inodesFree": 41766334,
      "inodes": 41942000,
      "inodesUsed": 1
     },
     "swap": {
      "time": "2024-12-19T11:30:04Z",
      "swapAvailableBytes": 0,
      "swapUsageBytes": 0
     }
    }
   ],
   "cpu": {
    "time": "2024-12-19T11:29:52Z",
    "usageNanoCores": 3449322665,
    "usageCoreNanoSeconds": 140384807847
   },
   "memory": {
    "time": "2024-12-19T11:29:52Z",
    "availableBytes": 5956784128,
    "usageBytes": 3958640640,
    "workingSetBytes": 1559408640,
    "rssBytes": 1343029248,
    "pageFaults": 0,
    "majorPageFaults": 0
   },
   "network": {
    "time": "2024-12-19T11:30:00Z",
    "name": "eth0",
    "rxBytes": 978602864,
    "rxErrors": 0,
    "txBytes": 1167096,
    "txErrors": 0,
    "interfaces": [
     {
      "name": "eth0",
      "rxBytes": 978602864,
      "rxErrors": 0,
      "txBytes": 1167096,
      "txErrors": 0
     }
    ]
   },
   "volume": [
    {
     "time": "2024-12-19T11:28:24Z",
     "availableBytes": 73673723904,
     "capacityBytes": 85886742528,
     "usedBytes": 0,
     "inodesFree": 41709102,
     "inodes": 41942000,
     "inodesUsed": 1,
     "name": "scripts"
    },
    {
     "time": "2024-12-19T11:28:24Z",
     "availableBytes": 73673723904,
     "capacityBytes": 85886742528,
     "usedBytes": 0,
     "inodesFree": 41709102,
     "inodes": 41942000,
     "inodesUsed": 2,
     "name": "logs"
    },
    {
     "time": "2024-12-19T11:28:24Z",
     "availableBytes": 73673723904,
     "capacityBytes": 85886742528,
     "usedBytes": 0,
     "inodesFree": 41709102,
     "inodes": 41942000,
     "inodesUsed": 1,
     "name": "repo"
    },
    {
     "time": "2024-12-19T11:28:24Z",
     "availableBytes": 7516180480,
     "capacityBytes": 7516192768,
     "usedBytes": 12288,
     "inodesFree": 4062379,
     "inodes": 4062388,
     "inodesUsed": 9,
     "name": "kube-api-access-xrhdf"
    }
   ],
   "ephemeral-storage": {
    "time": "2024-12-19T11:30:04Z",
    "availableBytes": 76115410944,
    "capacityBytes": 85886742528,
    "usedBytes": 973582336,
    "inodesFree": 41766334,
    "inodes": 41942000,
    "inodesUsed": 52
   },
   "process_stats": {
    "process_count": 0
   },
   "swap": {
    "time": "2024-12-19T11:29:52Z",
    "swapUsageBytes": 0
   }
  }
 ]
 }
}
```

**Comment by flphvlck** [comment]
I also bumped this issue. Similar scenario - short lived pods.

**Comment by jmcgrath207** [comment]
Hey y'all, just had a PR from @AverageMarcus to fix this.

https://github.com/jmcgrath207/k8s-ephemeral-storage-metrics/pull/168

It's not enabled by default for now, but you'll be able to enable it this way.
```yaml
metrics:
  # -- Enable garbage collection for metrics
  gc_enabled: true
  # -- The interval, in minutes, to perform garbage collection
  gc_interval: 5
```
As for why this is happening, I am assuming there are some issues with not receiving the Delete Event for the pod here in certain k8s cluster.
https://github.com/jmcgrath207/k8s-ephemeral-storage-metrics/blob/24744c33afa7b34bbd99b00bf03cd3376927f949/pkg/pod/k8s.go#L74-L79

Regardless, the GC feature here will serve as a reliable backup if synchronization issues arise.

If this works for y'all, I'll set the GC as the default option in upcoming releases.

**Comment by AverageMarcus** [comment]
Unfortunately I'm still missing some memory increase that only seems to be seen in our prod environment (likely due to the size of the cluster) so I'm going to get some profiling added in and see if I can pin-point exactly where the memory us building up.

<img width="1690" height="421" alt="Image" src="https://github.com/user-attachments/assets/609f7e13-0efe-4f11-8920-7dc123a7ab43" />

**Comment by AverageMarcus** [comment]
Been running the latest version with GC enabled for the past few days now in our production environment without any issues! 🎉 

This shows the before and after of enabling GC and the impact on "stale" metrics. (In the before, we triggered a pod restart every 24 hours)
<img width="1685" height="416" alt="Image" src="https://github.com/user-attachments/assets/f2fed330-28c2-4f63-8020-6ea2d1177d50" />

This shows the memory usage of k8s-ephemeral-storage-metrics pod (the spikes are pod starts). The GC was enabled ~12pm on the 5th and you can see a more level memory usage following that compared to the inclines seen before.

<img width="1685" height="416" alt="Image" src="https://github.com/user-attachments/assets/0e55ca07-2eb4-4792-8de9-e14618cf7c3f" />


@tomfankhaenel would be good to see if enabling GC in the latest version resolves your initial issues too 🤞

**Comment by jmcgrath207** [comment]
Closing as resolved (opt-in).

Garbage collection for stale pod/node metrics landed in #168, with memory and batching improvements in #172 and #175. The GC loop (`gcMetrics` in `pkg/pod/factory.go` and `pkg/node/factory.go`) evicts metrics for pods/nodes that no longer exist.

Note: GC ships **off by default**. To enable, set in your values:
```yaml
metrics:
  gc_enabled: true
  gc_interval: 5      # minutes
  gc_batch_size: 500
```

With GC on, vanished pods will no longer keep exporting their last value.

**Comment by jmcgrath207** [comment]
Closing as fixed by #168 / #172 / #175 (garbage collection). Enable via metrics.gc_enabled: true.
